README.md
# EE450 Socket Programming Project

## Personal Details

- Full Name: Hao Liang
- Student ID: 4773-3404-69

## Project Description

In this assignment, I implemented a library registration system based on TCP and UDP. I also implement the optional part (admin).

### Backend Server

There are three backend server: Server S, Server L and Server H. They managing different categories of books.

| Server   | Function                  | Input File     |
| -------- | ------------------------- | -------------- |
| Server S | Managing science books    | science.txt    |
| Server L | Managing literature books | literature.txt |
| Server H | Managing history books    | history.txt    |

### Main Server

There is one main server: `Server M`. It has two main functions:

1. The main server accepts login requests from the client and determines whether the user meets the login criteria based on encrypted user list information.
2. The main server accepts query requests from the client and sends requests to different backend servers based on specific requests to obtain book information. Then, it transmits the book information back to the client.

### Client

There is one client: `client`.

#### User Types and Privileges

#### Client Login
- **Initial Step**: Clients must enter a username and password to log in.

#### Regular Users
- **Access**: After logging in, regular users can check the status of books.
- **Book Status**: Includes "Available", "Not Available", and "Not Found".
- **Limitation**: Cannot view the quantity of books.
- **Book Quantity**: Every query for an "available" book decreases its quantity by 1 on the backend server.

#### Administrator Users
- **Privileges**: Administrators have management rights.
- **Book Information**: Can view the quantity of each book.
- **Query Impact**: Their queries do not affect book quantities.

### Communication protocol architecture of the system



<img src="./assets/architecture.png" alt="protocol architecture of the system" style="zoom:50%;" />

> Image Source: EE450 Socket Programming Project.pdf

The Client communicates with server M using TCP. server M communicates with the three backend servers using UDP.

### Port numbers

| Process  | Dynamic Ports |          Static Ports          |
| :------: | :-----------: | :----------------------------: |
| Server M |       -       | 1 UDP, 44469<br />1 TCP, 45469 |
| Server S |       -       |          1 UDP, 41469          |
| Server L |       -       |          1 UDP, 42469          |
| Server H |       -       |          1 UDP, 43469          |
|  Client  |     1 TCP     |    Dynamic port assignment     |

## Code Files

1. `client.cpp`: Client-side application to access the registration system and encrypt login information.

2. `serverM.cpp`: Main server that verifies member identity and coordinates with backend servers.

3. `serverS.cpp`: Backend server for the Science department, managing science book inventory.

4. `serverL.cpp`: Backend server for the Literature department, managing literature book inventory.

5. `serverH.cpp`: Backend server for the History department, managing history book inventory.

6. `build_encryption_map.h`: The encryption module in the project consists of two primary functions, `buildEncryptionMap()` and `encrypt()`. The encryption module is designed for secure data handling, including encrypting usernames and passwords for client-server communication in the system. After user type in their username and password, `client.cpp` will use this function to encrypt the information before sending them to the main server.

   1. `buildEncryptionMap()` function

      - Constructs and returns an unordered map (`std::unordered_map<char, char>`) for character-based encryption.

      - Encryption for lowercase and uppercase letters: Shifts each letter by 5 positions in the alphabet. Wraps around after 'z' for lowercase and 'Z' for uppercase.
      - Encryption for digits: Shifts each digit ('0'-'9') by 5 positions, with a wraparound after '9'.
      - Special characters remain unchanged.

   2. `encrypt()` Function

      - Parameters: Accepts a string (`std::string`) and the encryption map from `buildEncryptionMap()`.
      - Iterates through each character of the input string, encrypts it using the map, and constructs the encrypted string.
      - Leaves characters not in the encryption map (like special characters) unchanged. 


## Message Format

### Input file format

#### 1. member.txt for Server M
- **Format**: Each line contains a username and a password, separated by a comma and a space.
- **Username**: Alphanumeric, only lower case letters.
- **Password**: A mix of letters, numbers, and special characters.
- **Example Entry**: `otms, Uf@$b5wi` - 'otms' is the username, and 'Uf@$b5wi' is the password.

#### 2. literature.txt for Server L
- **Format**: Each line lists a literature book code and its available quantity, separated by a comma and a space.
- **Book Code**: Prefixed with 'L', followed by a numeric code.
- **Quantity**: Number of available copies.
- **Example Entry**: `L102, 7` - Book code 'L102' with 7 copies available.

#### 3. history.txt for Server H
- **Format**: Similar to literature.txt, but for history books.
- **Book Code**: Prefixed with 'H', followed by a numeric code.
- **Example Entry**: `H305, 0` - Book code 'H305' with 0 copies available.

#### 4. science.txt for Server S
- **Format**: Similar to literature.txt and history.txt, but for science books.
- **Book Code**: Prefixed with 'S', followed by a numeric code.
- **Example Entry**: `S305, 4` - Book code 'S305' with 4 copies available.

### Server M to Backend Server

### Client to Server M

1. ```C++
   // encrypted username and password, delimited by a space.
   string message = encryptedUsername + " " + encryptedPassword;
   strncpy(sendBuffer, message.c_str(), sizeof(sendBuffer));
   int byteSent = send(clientSocket, sendBuffer, sizeof(sendBuffer), 0);
   ```

2. ```C++
   // book code, the same as use input
   cin >> bookCode;
   strncpy(sendBuffer, bookCode.c_str(), sizeof(sendBuffer));
   ```

### Server M to Client

```C++
int responseType = 0;
// 1: authorize success, 2: password not match, 3: username not exist
// 4: request not match 5: available, 6: not available 7: not found
// 8. not found admin 9. available admin

switch (responseType) {
    case 1:
        strcpy(TCP_send_buffer, "Authorize");
        break;
    case 2:
        strcpy(TCP_send_buffer, "Unauthorized");
        break;
    case 3:
        strcpy(TCP_send_buffer, "NoUser");
        break;
    case 4:
        strcpy(TCP_send_buffer, "NoServer");
        break;
    case 5:
        strcpy(TCP_send_buffer, "Available");
        break;
    case 6:
        strcpy(TCP_send_buffer, "Unavailable");
        break;
    case 7:
        strcpy(TCP_send_buffer, "NoRecord");
        break;
    case 8:
        strcpy(TCP_send_buffer, "NoRecord");
        break;
    case 9:
        strcpy(TCP_send_buffer, UDP_receive_buffer);
        break;
    default:
        break;
}
int TCP_byteSent = send(acceptTCP_Socket, TCP_send_buffer, 200, 0);
```

In case 9: Server M receive the number of the book from the backend server. So `TCP_send_buffer` and `UDP_receive_buffer` is just the number of the certain book required by the client.

### Server M to Backend Server S, L, H

1. Book code query

   ```C++
   if (isAdmin) {
       TCP_receive_buffer[TCP_byteReceived - 2] = 'A';
       TCP_receive_buffer[TCP_byteReceived - 1] = '\0';
   }
                       
   int byteSent = sendto(UDP_Socket, TCP_receive_buffer, 200, 0, (struct sockaddr*)&targetServer, sizeof(targetServer));
   ```

   If the user is admin, then I add `A` at the send of the message. `TCP_receive_buffer` is the book code received from the client.

### Backend Server S, L, H to Server M

```C++
if (bookMap.find(bookCode) != bookMap.end()) {
    number = bookMap[bookCode];
    if (stoi(number) > 0 && !admin) {
        strcpy(sendBuffer, "The requested book is available.");

        // Update the number of books
        string new_number = to_string(stoi(number) - 1);
        bookMap[bookCode] = new_number;
        ofstream outFile;
        outFile.open("history.txt", ios::out);
        if (!outFile.is_open()) {
            cerr << "Unable to open file history.txt for writing" << endl;
            return 1;
        }

        for (const auto& pair : bookMap) {
            outFile << pair.first << ", " << pair.second << "\n";
        }

        outFile.close();
    }
    // Admin request, just send the number of books if it is not negative
    else if (stoi(number) >= 0 && admin) {
        strcpy(sendBuffer, number.c_str());
    }
    else {
        strcpy(sendBuffer, "The requested book is not available.");
    }
}
else {
    strcpy(sendBuffer, "Not able to find the book.");
}
int sendBytes = sendto(udpSocket, sendBuffer, sizeof(sendBuffer), 0, (struct sockaddr*)&clientService, sizeof(clientService));
```

- If user is admin, send the number of book
- Otherwise, send the availability status of the book.

## Project Idiosyncrasies

This project has specific failure conditions under which it might not perform as expected.

1. If the hardcoded port is occupied, then the corresponding server port will not start. For example, if the port `45496` is occupied, which is the TCP port of the main server, then the main server cannot run.
2. Once the client is launched, if it is closed, the system will be unable to continue functioning, and the client will be unable to reconnect to the main server.

## Reused Code

1. In `serverH.cpp`: 49, I referred to [Getting std :: ifstream to handle LF, CR, and CRLF?](https://stackoverflow.com/questions/6089231/getting-std-ifstream-to-handle-lf-cr-and-crlf) to solve the problem of reading a file with a Windows line break at the end of the line.

   ```C++
   // Delete the possible \r in the end of the line, which is the case in Windows
   // reference: https://stackoverflow.com/questions/6089231/getting-std-ifstream-to-handle-lf-cr-and-crlf
   if (!line.empty() && line.back() == '\r') {
       line.pop_back();
   }
   ```

- In `serverH.cpp`: 58, I referred [cplusplus.com]([cplusplus.com/reference/string/string/find_first_not_of/](https://cplusplus.com/reference/string/string/find_first_not_of/)) to delete the possible space in the beginning of the number:

  ```C++
  // Delete the possible space in the beginning of the number
  // reference: https://cplusplus.com/reference/string/string/find_first_not_of/
  num.erase(0, num.find_first_not_of(" \n\r\t"));
  bookMap[code] = num;
  ```

  

history.txt
H305, 0
H308, 4
H207, 1
H201, 7
H103, 5
H104, 7


windows_serverM.cpp
//
// Created by Administrator on 2023/11/15.
//
#include <iostream>
#include <winsock2.h>
#include <ws2tcpip.h>
#include <tchar.h>
#include <unordered_map>
#include <string>
#include <fstream>
#include <sstream>
using namespace std;

int main(int argc, char* argv[]) {
    
    unordered_map<string, int> portNumbers;
    portNumbers["S"] = 41469;
    portNumbers["L"] = 42469;
    portNumbers["H"] = 43469;
    
    bool isAuthorized = false;
    bool isAdmin = false;
    char UDP_receive_buffer[200] = {0};
    char bookCode[200] = {0};
    
    // ================== Step 0: Print out the message and load member data ==================
    
    cout <<"\nMain Server is up and running\n" << endl;
    
    unordered_map<string, string> members;
    
    ifstream file;
    file.open("member.txt", ios::in);
    if (!file.is_open()) {
        cout << "Error opening member.txt" << endl;
        return 1;
    }
    
    string line;
    while (getline(file, line)) {
        istringstream iss(line);
        string username, password;
        if (getline(iss, username, ',') && getline(iss, password)) {
            // Delete the possible space in the beginning of the password
            password.erase(0, password.find_first_not_of(" \n\r\t"));
            
            members[username] = password;
        }
    }
    
    file.close();
    
    cout << "Main Server loaded member list." << endl;
    
    // ================== Step 1: Set up DLL ==================
    
    // Initialize Winsock
    WSADATA wsaData;
    int wsaErr;
    WORD wVersionRequested = MAKEWORD(2, 2);
    wsaErr = WSAStartup(wVersionRequested, &wsaData);
    if (wsaErr != 0) {
        cout << "WSAStartup failed with error: " << wsaErr << endl;
        return 0;
    }
    else {
//        cout << "WSAStartup() is OK!" << endl;
//        cout << "The status: " << wsaData.szSystemStatus << endl;
    }
    
    // ================== Step 2: Set up the TCP and UDP Socket ==================
    
    // TCP socket
    SOCKET TCP_Socket;
    int TCP_Port = 45469;
    TCP_Socket = INVALID_SOCKET;
    TCP_Socket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (TCP_Socket == INVALID_SOCKET) {
        cout << "TCP socket() failed with error: " << WSAGetLastError() << endl;
        WSACleanup();
        return 0;
    }
    else {
//        cout << "TCP socket() is OK!" << endl;
    }
    
    // UDP socket
    SOCKET UDP_Socket;
    int UDP_Port = 44469;
    UDP_Socket = INVALID_SOCKET;
    UDP_Socket = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
    if (UDP_Socket == INVALID_SOCKET) {
        cout << "UDP socket() failed with error: " << WSAGetLastError() << endl;
        WSACleanup();
        return 0;
    }
    else {
//        cout << "UDP socket() is OK!" << endl;
    }
    
    // ================== Step 3: Bind the TCP and UDP Socket ==================
    
    // TCP bind
    sockaddr_in clientService;
    clientService.sin_family = AF_INET;
    InetPton(AF_INET, _T("127.0.0.1"), &clientService.sin_addr.s_addr);
    clientService.sin_port = htons(TCP_Port);
    if (bind(TCP_Socket, (SOCKADDR*)&clientService, sizeof(clientService)) == SOCKET_ERROR) {
        cout << "TCP bind() failed with error: " << WSAGetLastError() << endl;
        closesocket(TCP_Socket);
        WSACleanup();
        return 0;
    }
    else {
//        cout << "\nTCP bind() is OK!\n" << endl;
    }
    
    // UDP bind
    sockaddr_in backendService;
    backendService.sin_family = AF_INET;
    InetPton(AF_INET, _T("127.0.0.1"), &backendService.sin_addr.s_addr);
    backendService.sin_port = htons(UDP_Port);
    if (bind(UDP_Socket, (SOCKADDR*)&backendService, sizeof(backendService)) == SOCKET_ERROR) {
        cout << "UDP bind() failed with error: " << WSAGetLastError() << endl;
        closesocket(UDP_Socket);
        WSACleanup();
        return 0;
    }
    else {
//        cout << "\nUDP bind() is OK!\n" << endl;
    }
    
    // ================== Step 4: Listen on the TCP Socket ==================
    if (listen(TCP_Socket, 1) == SOCKET_ERROR) {
        cout << "listen() failed with error: " << WSAGetLastError() << endl;
        closesocket(TCP_Socket);
        WSACleanup();
        return 0;
    }
    else {
//        cout << "listen() is OK! Waiting for connections!" << endl;
    }
    
    
    // ================== Step 5: Accept a Client Socket ==================
    SOCKET acceptTCP_Socket = accept(TCP_Socket, NULL, NULL);
    if (acceptTCP_Socket == INVALID_SOCKET) {
        cout << "accept() failed with error: " << WSAGetLastError() << endl;
        closesocket(TCP_Socket);
        WSACleanup();
        return -1;
    }
    else {
//        cout << "Accepted connection" << endl;
    }
    
    while (true) {
        // ================== Step 6: Chat with the client ==================
        char TCP_receive_buffer[200] = {0};
        int responseType = 0;
        // 1: authorize success, 2: password not match, 3: username not exist
        // 4: request not match 5: available, 6: not available 7: not found
        // 8. not found admin 9. available admin
        
        int TCP_byteReceived = recv(acceptTCP_Socket, TCP_receive_buffer, 200, 0);
        if (TCP_byteReceived > 0) {
            // ================== Step 7: Authorize ==================
            if (!isAuthorized) {
                string username, password;
                string bufferStr(TCP_receive_buffer);
                istringstream iss(bufferStr);
                iss >> username >> password;
                cout << "Main Server received the username and password from the client using TCP over port " << TCP_Port
                   << "."  << endl;
                // If the username exists in the member.txt
                if (members.find(username) != members.end()) {
                    // If the password matches the username
                    if (members[username] == password) {
                        isAuthorized = true;
                        
                        // Admin request
                        if (username == "firns" && password == members["firns"]) {
                            isAdmin = true;
                        }
                        cout << "Password " << password << " matches the username. Send a reply to the client." << endl;
                        responseType = 1;
                    }
                        // If the password does not match the username
                    else {
                        cout << "Password " << password << " does not match the username. Send a reply to the client." << endl;
                        responseType = 2;
                    }
                }
                    // If the username does not exist in the member.txt
                else {
                    cout << username << " is not registered. Send a reply to the client." << endl;
                    responseType = 3;
                }
            } else {
                strcpy(bookCode, TCP_receive_buffer);
                // ================== Step 8: Send the request to the backend server ==================
                cout << "Main Server received the book request from the client using TCP over port " << TCP_Port << "." << endl;
                string target(1, TCP_receive_buffer[0]);
                
                if (portNumbers.find(target) == portNumbers.end()) {
                    cout << "Did not found " << TCP_receive_buffer << " in the book code list." << endl;
                    responseType = 4;
                }
                else {
                    int targetPort = portNumbers[target];
                    cout << "Found " << TCP_receive_buffer << " located at Server " << target << ". Send to Server " << target << "." << endl;
                    sockaddr_in targetServer;
                    targetServer.sin_family = AF_INET;
                    InetPton(AF_INET, _T("127.0.0.1"), &targetServer.sin_addr.s_addr);
                    targetServer.sin_port = htons(targetPort);
                    
                    // Admin request to backend server
                    if (isAdmin) {
                        TCP_receive_buffer[TCP_byteReceived - 2] = 'A';
                        TCP_receive_buffer[TCP_byteReceived + 2] = '\0';
                    }
                    
                    int byteSent = sendto(UDP_Socket, TCP_receive_buffer, 200, 0, (SOCKADDR*)&targetServer,
                                          sizeof(targetServer));
                    
                    if (byteSent > 0) {
                        // ================== Step 9: Receive the response from the backend server ==================
                        int byteRecv = recvfrom(UDP_Socket, UDP_receive_buffer, 200, 0, NULL, NULL);
                        if (byteRecv > 0) {
                            cout << "Main Server received from Server " << target << " the book status result using UDP over port " << UDP_Port << "." << endl;
                            if (strcmp(UDP_receive_buffer, "The requested book is available.") == 0 && !isAdmin) {
                                cout << "The requested book " << bookCode << " is available." << endl;
                                responseType = 5;
                            }
                            else if (strcmp(UDP_receive_buffer, "The requested book is not available.") == 0) {
                                cout << "The requested book " << bookCode << " is NOT available." << endl;
                                responseType = 6;
                            }
                            
                            else if (strcmp(UDP_receive_buffer, "Not able to find the book.") == 0 && !isAdmin) {
                                cout << "Not able to find the book " << bookCode << " in the system." << endl;
                                responseType = 7;
                            }
                            else if (strcmp(UDP_receive_buffer, "Not able to find the book.") == 0 && isAdmin) {
                                cout << "Not able to find the book " << bookCode << " in the system." << endl;
                                responseType = 8;
                            }
                            else if (isAdmin) {
                                responseType = 9;
                            }
                        }
                        else {
                            cout << "recvfrom() backend server failed with error: " << WSAGetLastError() << endl;
                            WSACleanup();
                            return 0;
                        }
                    }
                    else {
                        cout << "sendto() backend server failed with error: " << WSAGetLastError() << endl;
                        WSACleanup();
                        return 0;
                    }
                }
                
            }
            
            // ================== Step 10: Send the response to the client ==================
            char TCP_send_buffer[200] = {0};
            switch (responseType) {
                case 1:
                    strcpy_s(TCP_send_buffer, "Authorize");
                    break;
                case 2:
                    strcpy_s(TCP_send_buffer, "Unauthorized");
                    break;
                case 3:
                    strcpy_s(TCP_send_buffer, "NoUser");
                    break;
                case 4:
                    strcpy_s(TCP_send_buffer, "NoServer");
                    break;
                case 5:
                    strcpy_s(TCP_send_buffer, "Available");
                    break;
                case 6:
                    strcpy_s(TCP_send_buffer, "Unavailable");
                    break;
                case 7:
                    strcpy_s(TCP_send_buffer, "NoRecord");
                    break;
                case 8:
                    strcpy_s(TCP_send_buffer, "NoRecord");
                    break;
                case 9:
                    strcpy_s(TCP_send_buffer, UDP_receive_buffer);
                    break;
                default:
                    break;
            }
            int TCP_byteSent = send(acceptTCP_Socket, TCP_send_buffer, 200, 0);
            if (TCP_byteSent > 0) {
                if (isAuthorized && responseType >= 4 && responseType <= 8) {
                    cout << "Main Server sent the book result to the client." << endl;
                }
                else if (responseType == 9) {
                    cout << "Number of books " << bookCode << " available is: " << TCP_send_buffer[0] << "." << endl;
                    cout << "Main Server sent the book result to the client." << endl;
                }
            }
            else {
                cout << "send() failed with error: " << WSAGetLastError() << endl;
                closesocket(TCP_Socket);
                WSACleanup();
                return 0;
            }
        }
    }
}